components/paging/pager-load-more.js

```diff
@@ -0,0 +1,146 @@
+			/**
+			 * Total number of items. If not specified, neither it nor the count of items showing will be displayed.
+			 * @type {number}
+			 */
+			itemCount: { type: Number, attribute: 'item-count', reflect: true },
+			/**
+			 * The number of additional items to load.
+			 * @type {number}
+			 */
+			pageSize: { type: Number, attribute: 'page-size', reflect: true },
+			/**
+			 * The number of items showing. Assigned by PageableMixin.
+			 * @ignore
+			 */
+			itemShowingCount: { attribute: false, type: Number },
+			_loading: { state: true }
+		};
+	}
+
+	static get styles() {
+		return [ buttonStyles, labelStyles, offscreenStyles, css`
+			:host {
+				display: block;
+			}
+			:host(:not([has-more])),
+			:host([hidden]) {
+				display: none;
+			}
+			button {
+				align-items: center;
+				background-color: var(--d2l-color-regolith);
+				border: 1px solid var(--d2l-color-sylvite);
+				display: flex;
+				gap: 0.5rem;
+				justify-content: center;
+				width: 100%;
+			}
+			button:hover {
+				background-color: var(--d2l-color-sylvite);
+				border-color: var(--d2l-color-gypsum);
+			}
+			.action {
+				color: var(--d2l-color-celestine);
+			}
+			.separator {
+				border-right: 1px solid var(--d2l-color-mica);
+				height: 0.8rem;
+			}
+			.info {
+				color: var(--d2l-color-galena);
+				font-weight: 400;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+		this.hasMore = false;
+		this.itemCount = -1;
+		/** @ignore */
+		this.itemShowingCount = 0;
+		this.pageSize = 50;
+		this._loading = false;
+	}
+
+	static get focusElementSelector() {
+		return 'button';
+	}
+
+	render() {
+		if (!this.hasMore) return;
+		return html`<button ?disabled="${this._loading}" class="d2l-label-text" @click="${this._handleClick}" type="button">
+			${this._loading ? html`
+				<span class="d2l-offscreen" role="alert">${this.localize('components.pager-load-more.status-loading')}</span>
+				<d2l-loading-spinner size="24"></d2l-loading-spinner>
+			` : html`
+				<span class="action">${this.localize('components.pager-load-more.action', { count: this.pageSize })}</span>
+				${this.itemCount > -1 ? html`
+					<span class="d2l-offscreen">${getSeparator({ nonBreaking: true })}</span>
+					<span class="separator"></span>
+					<span class="info">${this.localize('components.pager-load-more.info', { showingCount: this.itemShowingCount, totalCount: this.itemCount })}</span>
+				` : nothing}
+			`}
+		</button>
+		`;
+	}
+
+	async _handleClick() {
+		const pageable = findComposedAncestor(this, node => node._pageable);
+		if (!pageable) return;
+		const lastItemIndex = pageable._getLastItemIndex();
+
+		await new Promise(resolve => {
+			this._loading = true;
+			this.dispatchEvent(new CustomEvent('d2l-pager-load-more', {
+				bubbles: false,
+				composed: false,
+				detail: { complete: resolve }
+			}));
+		});
+		this._loading = false;
+
+		const item = pageable._getItemByIndex(lastItemIndex + 1);
```

Comment:
Is there any concern that when they resolve the promise that their collection hasn't actually re-rendered yet and then this line wouldn't find the new item(s)?

Seems like the typical workflow will be for the consumer to do a `fetch` call, get some new data, update a local collection and then resolve this promise. But then their re-render could easily happen afterwards. 🤔

Reply:
Possibly yeah. Perhaps it would be best practice that they would `await this.updateComplete;` before calling `complete()`.  Even at that point, it is possible that the list-item itself has not been updated, but `d2l-pager-load-more` will wait for it's `updateComplete` if defined.  At that point it should be ready to be focused... unless... the item isn't draggable, isn't selectable, doesn't define a primary action (navigation or button), doesn't provide any secondary actions, and is also a hypermedia component that is yet to load and render. I'm not sure how likely it is to have that scenario... but if we did, maybe we'd wait some amount of time or check on an interval for a focusable element. Thoughts?

Reply:
What if `PageableMixin` provided a way for the pager to register with it to be notified whenever new items were added? I guess List doesn't actually know that though eh... I was thinking the items fired an event that List was listening for but that's just the selection observer stuff. To finish my thought: if such a thing existed, the pager could just wait for a new item to be added _after_ it had been put into "load more" state.

Reply:
Yeah I was thinking about that a bit. Items can be added to the list for a couple different reasons - via drag & drop or by bulk actions, and the list doesn't really know why they were added.  We could require the items be added with a special attribute to identify them, but it would require removing that attribute, probably outside of render.  Not sure if I like that any better.

Reply:
For sure it's definitely not straightforward. We could block adding while load-more is in progress or something, but yeah... complex.